source/slang/slang-check-decl.cpp

```diff
@@ -9753,6 +9804,8 @@ namespace Slang
                     diagnoseUndeclaredCapability(funcDecl, Diagnostics::useOfUndeclaredCapability, failedAvailableCapabilityConjunction);
                     funcDecl->inferredCapabilityRequirements = declaredCaps;
                 }
+                else
+                    funcDecl->inferredCapabilityRequirements.simpleJoinWithSetMask(declaredCaps, CapabilityName::stage);
```

Comment:
Can you explain why  this is needed with a concrete example?

Reply:
the general problem here will be described with a concrete example below: 
1. I assign capability to a function, this resolves as a `declaredCaps`. 
2. 2. We assign a shader stage to `declaredCaps`. 
3. 3. Now when we call this function, since we likely never join `declaredCaps` with `inferredCapability`, future checks of `inferredCapability` will not have data of any shader stage requirement.

concrete example:
```
[ForceInline]
[require(glsl_hlsl_spirv, shadermemorycontrol_compute)]
public void memoryBarrierAtomicCounter()
{
    __target_switch
    {
    case hlsl: __intrinsic_asm "AllMemoryBarrier()";
    case glsl: __intrinsic_asm "memoryBarrierBuffer()";
    case spirv:
    {
        spirv_asm
        {
            OpCapability Shader;
            OpMemoryBarrier Device AcquireRelease|UniformMemory
        };
    }
    }
}
```
If we try to use this function in a vertex shader without the change provided, the code will not error on the Slang side of things since the shader does not propagate `compute` into the "inferredCapabilityRequirements".

Technically speaking, "join" is more correct that joinWithSetMask, but there are issues with doing this on the 
design & performance side of things with the current state of the capabilities system and how we plan to auto-magically manage code based on capabilities.